Implement the `bootc` provision plugin

ckyrouac opened the pull request:
This creates a new provision plugin that is built on top of the existing TestCloud (virtual) plugin. It adds new parameters to pass a Containerfile or container image. The plugin will then build a container image (if necessary) then build a bootc disk image from the container image using bootc image builder. Currently, bootc requires podman to be run as root when building a disk image. This is typically handled by running a podman machine as root.

An additional parameter "add-deps" toggles building a derived container image with the tmt test requirements.

Resolves #3013 

Pull Request Checklist

* [X] implement the feature
* [x] write the documentation
* [x] extend the test coverage
* [x] adjust plugin docstring
* [x] modify the json schema
* [x] include a release note

---
This is a work in progress. I'm opening this PR early to get feedback on the high level design. I will add tests, docs, etc. after we solidify the higher level design.

If you want to try running the code here is an example fmf plan:

```
provision:
  how: bootc
  containerimage: quay.io/centos-bootc/centos-bootc:stream9
  disk: 20
summary: Testing bootc plugin
execute:
  how: tmt
  script: |
    echo "ok"
```

ckyrouac commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,157 @@
+
+    # build the "base" or user supplied container image
+    def _build_base_image(self, containerfile: str, workdir: str) -> str:
+        image_tag = "localhost/tmtbase"  # TODO: how do I get a unique ID here?
```
I think using the tmt run # will work for this. Is there a good way to access it here? Or is there another unique ID I could use for the image tag?

happz replied:
How unique does the "unique ID" need to be? Do you wish it to be the same when the user re-runs the previous run, or should it be unique every time `tmt run` is invoked?

ckyrouac replied:
I think keeping the ID the same when re-running the previous run will work. If I understand how tmt runs work, this will allow modifying the test code without rebuilding the bootc disk which can take a few minutes.

ckyrouac commented:
@cgwalters fyi

ckyrouac commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,157 @@
+    _guest = None
+
+    # build a "modified" container image from the base image with tmt dependencies added
+    def _build_derived_image(self, base_image: str) -> str:
```
I went this route instead of using podman-bootc (to leverage smbios) primarily because podman-bootc is not yet released in the official repos (it's in [German's copr repo](https://github.com/containers/podman-bootc?tab=readme-ov-file#fedora)).

ckyrouac commented:
@cgwalters fyi

cgwalters commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,157 @@
+
+            RUN \
+            dnf -y install cloud-init rsync && \
+            ln -s ../cloud-init.target /usr/lib/systemd/system/default.target.wants && \
```
Minor nit, missing the `dnf clean all` so we're leaking unnecessary cruft into `/var`; to make this sustainable long term we probably need to go to a declarative input tool, cc https://gitlab.com/fedora/bootc/osbuild-cfg

ckyrouac replied:
good catch, updated in latest push

cgwalters commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,157 @@
+            "-v", "/var/lib/containers/storage:/var/lib/containers/storage",
+            "--security-opt", "label=type:unconfined_t",
+            "-v", f"{self.workdir}:/output",
+            "quay.io/centos-bootc/bootc-image-builder:latest", "build",
```
I've been trying to avoid people hardcoding this because we already have the situation where there are two builders (one upstream, one for RHEL). Exact same comment as https://github.com/containers/podman-bootc/pull/58#discussion_r1668661289

It's actually somewhat likely that we move out of the quay.io/centos-bootc namespace at some point; while we'd probably maintain a redirect for "a while", with the "consume a label" approach it automatically derives from the base image which avoids hardcoding entirely.

happz replied:
Would it make sense to make the namespace configurable? In any case, having a default as a module-level constant should happen, but is it something the user might wish to override? E.g. `namespace: quay.io/acme/bootc`? I think it would be good to open this piece to users.

ckyrouac replied:
added a parameter for the user to configure this in the latest push

cgwalters commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,157 @@
+            RUN \
+            dnf -y install cloud-init rsync && \
+            ln -s ../cloud-init.target /usr/lib/systemd/system/default.target.wants && \
+            rm /usr/local -rf && ln -sr /var/usrlocal /usr/local && mkdir -p /var/usrlocal/bin
```
We should also probably add a TODO somewhere for tmt to put its temporary binaries somewhere other than `/usr/local` too.

happz replied:
I believe it might be this issue, spawned by slightly different conditions, but maybe hitting the same wall? https://github.com/teemtee/tmt/issues/2884

cgwalters reviewed (commented):
Mostly a skim (I don't know the tmt codebase either) but looks sane to me!

happz commented:
* LGTM, I like the reuse of existing plugins, that's very nice.
* Would it also make sense to support the `container` plugin? IIUIC, the image produced by `podman` can be run both with `podman run`, or, converted to qcow2, as a VM. I could imagine both ways being available to use, maybe through some `runtime-plugin: container|virtual` switch. Is it a bad idea?
* I will have more low-level comments related to the actual implementation, but I won't bother you now till you receive the high-level review. Just ping when you're ready for the boring stuff :)

martinhoyer reviewed (commented):
I'm still trying to speedrun reading/learning all things bootc (and tmt provisioning plugins), but fwiw, looks cool to me.  

@happz about the container plugin support - [Fedora docs](https://docs.fedoraproject.org/en-US/bootc/getting-started/#_local_development) says:  

> for fully-fledged tests it is not recommended to run a bootable container via, for instance, podman-run. One reason among others is that the filesystem is writable when being executed as an OCI container while most of the filesystem is mounted read-only on a deployed bootc system. That means the running container behaves differently than a deployed system. Yet, if you desire to run some quick tests it is recommended to run the container in detached mode.  

 From what I understand, `podman-bootc` could be pretty cool to use, once available.

happz commented:
> I'm still trying to speedrun reading/learning all things bootc (and tmt provisioning plugins), but fwiw, looks cool to me.
> 
> @happz about the container plugin support - [Fedora docs](https://docs.fedoraproject.org/en-US/bootc/getting-started/#_local_development) says:
> 
> > for fully-fledged tests it is not recommended to run a bootable container via, for instance, podman-run. One reason among others is that the filesystem is writable when being executed as an OCI container while most of the filesystem is mounted read-only on a deployed bootc system. That means the running container behaves differently than a deployed system. Yet, if you desire to run some quick tests it is recommended to run the container in detached mode.
> 
> From what I understand, `podman-bootc` could be pretty cool to use, once available.

I agree, it's not a perfect 1:1 substitution, but, exactly: for quick tests or basic test development, it may give me results faster than VM. I for one work on binutils and C/C++ toolchain in general, and my area of focus is fairly simple - compile this, run `objdump` on that, grep for expected section names, this kind of stuff. Learning about a typo in my reproducer quickly is very valuable, and once I'm done, I can always use full VM. I do it today already: I develop tests with `container`, then I switch to `beaker` or `1minutetip` to get a more real environment before committing the new test to git. So, `podman-run`, for my trivial component, would be very welcome, even with caveats like the one you mentioned :)

martinhoyer commented:
> I agree, it's not a perfect 1:1 substitution, but, exactly: for quick tests or basic test development, it may give me results faster than VM. I for one work on binutils and C/C++ toolchain in general, and my area of focus is fairly simple - compile this, run `objdump` on that, grep for expected section names, this kind of stuff. Learning about a typo in my reproducer quickly is very valuable, and once I'm done, I can always use full VM. I do it today already: I develop tests with `container`, then I switch to `beaker` or `1minutetip` to get a more real environment before committing the new test to git. So, `podman-run`, for my trivial component, would be very welcome, even with caveats like the one you mentioned :)

Thank you for the insight in your development process. I hope I can see it in more detail one day.

ckyrouac commented:
> Would it also make sense to support the container plugin? IIUIC, the image produced by podman can be run both with podman run, or, converted to qcow2, as a VM. I could imagine both ways being available to use, maybe through some runtime-plugin: container|virtual switch. Is it a bad idea?

I think the existing container plugin will handle this case without any additional code. The bootc image is just another container that can be run like a typical image.

lukaszachy reviewed (commented):
`--add-deps` doesn't install require/recommend yet, right? 

Either way, we have chicken&egg problem in the case 'dist-git-source' is used as the list of packages is known _after_ provision :/ But IMO we can ignore that for now to have something working.

happz commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,232 @@
+        ''')
+        containerfile = containerfile_template.substitute({"base_image": base_image})
+        with open(f'{self.workdir}/Containerfile', 'w') as file:
+            file.write(containerfile)
```
Could be simplified a bit, we're slowly transitioning to how `pathlib` lets us work with files, `${path}.write_text()` should work like a charm:

```python
(self.workdir / 'Containerfile').write_text(containerfile_template.substitute({"base_image": base_image}))
```

happz commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,232 @@
+            self._build_bootc_disk(containerimage, data.image_builder)
+        else:
+            self._logger.fail("Either containerfile or containerimage must be specified.")
+            raise SystemExit(1)
```
This is definitely worth an exception, e.g. `raise tmt.utils.ProvisionError("Either containerfile or containerimage must be specified.")`. tmt has some basic exception rendering, please leave the reporting and exit strategy to tmt core, it will be fine :)

happz commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,232 @@
+        data.set_image(f"file://{self.workdir}/qcow2/disk.qcow2")
+        data.show(verbose=self.verbosity_level, logger=self._logger)
+
+        containerimage = ""
```
Please, don't use an empty string to represent "unset" or "undefined" or "undecided" value. It's very hard for linters to distinguish from strings that have value, that are all the same to linters. `None` is a more easier to track and much better when it comes to report undefined/unset/undecided/unknown value is passed where defined/set/decided/known value is expected.

IIUIC, eventually this name will carry a valid value, either from `data.containerimage` or from `self._build_base_image()`, therefore I'd drop this declaration completely, `containerimage = data.containerimage` below should set it well-enough for both runtime and linters to understand it.

happz commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,232 @@
+    # build a "modified" container image from the base image with tmt dependencies added
+    def _build_derived_image(self, base_image: str) -> str:
+        self._logger.debug("Building modified container image with necessary tmt packages/config")
+        containerfile_template = Template('''
```
Please, use Jinja templates, it's the solution used throughout the tmt code. See https://github.com/teemtee/tmt/blob/main/tmt/utils/templates.py#L314

happz commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,232 @@
+    _data_class = BootcData
+    _guest_class = GuestTestcloud
+    _guest = None
+    _id = str(uuid.uuid4())[:8]
```
What is the purpose of this id? Would it be possible to use `Run.unique_id`, https://github.com/teemtee/tmt/blob/main/tmt/base.py#L3317? It should be available somewhere from `self.parent.step.my_run.unique_id`.

We should think twice before adding new kind ID, especially when it's with no clear connection to run owning images being built, might be much harder to locate them when removing things.

ckyrouac replied:
Thanks, I wasn't sure how to access the unique run id so I just generated one. It is appended to the container images tags to keep them unique for cleanup, etc. I switched to using `self.parent.step.my_run.unique_id`.

happz replied:
The attribute is no longer used, and it should be removed.

happz commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,232 @@
+             building the bootc disk image.
+             """)
+
+    def set_image(self, image: str) -> None:
```
This kind of setters is not necessary, `self.image = image` is perfectly acceptable.

happz commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,232 @@
+class BootcData(tmt.steps.provision.testcloud.ProvisionTestcloudData):
+    containerfile: Optional[str] = field(
+        default=None,
+        option=('--containerfile'),
```
`()` wrapping is not needed when there is only one option.

happz commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,232 @@
+            self.containerimage = containerimage
+
+    def remove(self) -> None:
+        if self.containerimage:
```
I'm not 100 % sure, but if `self.containerimage` was not assigned a value in `__init__()`, this test might fail because Python would report a missing attribute. It would be better to mark `containerimage` as `Optional[str]` and give it `None` as a default - it obviously can be irrelevant, `None` is easier for linters to track and prevent the use.

ckyrouac commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,232 @@
+                 name: Optional[str] = None,
+                 parent: Optional[tmt.utils.Common] = None,
+                 logger: tmt.log.Logger,
+                 containerimage: Optional[str]) -> None:
```
ah, yea that’s right. I had originally marked containerimage as optional because I had set `_guest_class = GuestBootc` which caused the `GuestBootc` class to be instantiated without a `containerimage`. Using `GuestBootc` as the `_guest_class` was causing some other odd issues (the tests would pass when running them individually, but when running multiple tests only the first would pass) so I flipped it back to GuestTestcloud which seems to work as expected.

happz replied:
I'm still convinced this can be improved: given the code in `go()`, `containerimage` will never be `None`. It will either be a string returned by `_build_derived_image()`, or a string returned by `_build_base_image()`, in any case `containerimage` variable in `go()` will always contain a valid string. Therefore this parameter does not need to be `Optional`, and the test below is also superfluous.

ckyrouac commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,232 @@
+        self._logger.debug("Building bootc disk image")
+        tmt.utils.Command(
+            "podman", "run", "--rm", "--privileged",
+            "-v", "/var/lib/containers/storage:/var/lib/containers/storage",
```
Moved the string to a constant at the top of the module. This is the default location for podman container storage. It can be changed in podman configs. I'm not sure it's worth adding an option to change it until we find it necessary. If you think we should add the option up front that works too.

happz replied:
Hm, I'm not familiar enough with `podman` & co. to tell how often one might need to change this path. I suppose it's safe for now to be marked as a "constant", we can always add an option later if it becomes clear there's a use case for it.

ckyrouac commented:
@happz thanks for the review! I believe I addressed all your suggestions.

ckyrouac commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,241 @@
+
+    def _get_id(self) -> str:
+        # FIXME: cast() - https://github.com/teemtee/tmt/issues/1372
+        parent = cast(tmt.steps.provision.Provision, self.parent)
```
copied this from https://github.com/teemtee/tmt/blob/21c9e9793f5c81a6546eb2da5abeec87efcc7685/tmt/steps/provision/__init__.py#L789

happz commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,241 @@
+from tmt.utils.templates import render_template
+
+DEFAULT_IMAGE_BUILDER = "quay.io/centos-bootc/bootc-image-builder:latest"
+CONTAINER_STORAGE_DIR = "/var/lib/containers/storage"
```
This should be `Path` instance right here and now. We try to wrap path-like strings with `tmt.utils.Path` as soon as they enter tmt internals.

happz commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,241 @@
+        """ Build a "derived" container image from the base image with tmt dependencies added """
+        if not self.workdir:
+            raise tmt.utils.ProvisionError(
+                "self.workdir must be defined")
```
To an end user, this error message is not very helpful. IIUIC, `assert self.workdir is not None  # narrow type` should be good enough, it's a common pattern that `self.workdir` remains `Optional`.

martinhoyer commented:
@happz let's rebase and run the pipeline here?

thrix commented:
We agreed to focus on getting this in 1.38, the due date for finishing is 24th October.

ckyrouac commented:
thanks for looking at this and planning for 1.38. Is there anything I can do to help move it along?

thrix commented:
> thanks for looking at this and planning for 1.38. Is there anything I can do to help move it along?

@ckyrouac not really, the team is overhelmed with reviews.

I am getting validation errors, I assume the jsonschema needs an update for the new provision plugin:

```
❯ pwd
/var/home/mvadkert/git/github.com/teemtee/tmt/tests/provision/bootc/data

❯ tmt lint
    warn: /containerfile_includes_deps:provision - {'how': 'bootc', 'add-deps': False, 'containerfile': '$TMT_BOOTC_CONTAINERFILE_RUNDIR/includes_deps.containerfile', 'containerfile-workdir': '.', 'disk': 20} is not valid under any of the given schemas
    warn: /containerfile_needs_deps:provision - {'how': 'bootc', 'add-deps': True, 'containerfile': '$TMT_BOOTC_CONTAINERFILE_RUNDIR/needs_deps.containerfile', 'containerfile-workdir': '.', 'disk': 20} is not valid under any of the given schemas
    warn: /image_includes_deps:provision - {'how': 'bootc', 'add-deps': False, 'containerimage': 'localhost/tmt-bootc-includes-deps', 'disk': 20} is not valid under any of the given schemas
    warn: /image_needs_deps:provision - {'how': 'bootc', 'add-deps': True, 'containerimage': 'localhost/tmt-bootc-needs-deps', 'disk': 20} is not valid under any of the given schemas
/containerfile_includes_deps
warn C000 key "add-deps" not recognized by schema /schemas/provision/container
warn C000 key "containerfile" not recognized by schema /schemas/provision/container
warn C000 key "containerfile-workdir" not recognized by schema /schemas/provision/container
warn C000 key "disk" not recognized by schema /schemas/provision/container
warn C000 value of "how" is not "container"
warn C000 key "add-deps" not recognized by schema /schemas/provision/mrack
warn C000 key "containerfile" not recognized by schema /schemas/provision/mrack
warn C000 key "containerfile-workdir" not recognized by schema /schemas/provision/mrack
warn C000 key "disk" not recognized by schema /schemas/provision/mrack
warn C000 value of "how" is not "beaker"
warn C000 key "add-deps" not recognized by schema /schemas/provision/virtual
warn C000 key "containerfile" not recognized by schema /schemas/provision/virtual
warn C000 key "containerfile-workdir" not recognized by schema /schemas/provision/virtual
warn C000 value of "how" is not "virtual"
warn C000 key "add-deps" not recognized by schema /schemas/provision/artemis
warn C000 key "containerfile" not recognized by schema /schemas/provision/artemis
warn C000 key "containerfile-workdir" not recognized by schema /schemas/provision/artemis
warn C000 key "disk" not recognized by schema /schemas/provision/artemis
warn C000 value of "how" is not "artemis"
warn C000 key "add-deps" not recognized by schema /schemas/provision/minute
warn C000 key "containerfile" not recognized by schema /schemas/provision/minute
warn C000 key "containerfile-workdir" not recognized by schema /schemas/provision/minute
warn C000 key "disk" not recognized by schema /schemas/provision/minute
warn C000 value of "how" is not "minute"
warn C000 key "add-deps" not recognized by schema /schemas/provision/connect
warn C000 key "containerfile" not recognized by schema /schemas/provision/connect
warn C000 key "containerfile-workdir" not recognized by schema /schemas/provision/connect
warn C000 key "disk" not recognized by schema /schemas/provision/connect
warn C000 value of "how" is not "connect"
warn C000 key "add-deps" not recognized by schema /schemas/provision/local
warn C000 key "containerfile" not recognized by schema /schemas/provision/local
warn C000 key "containerfile-workdir" not recognized by schema /schemas/provision/local
warn C000 key "disk" not recognized by schema /schemas/provision/local
warn C000 value of "how" is not "local"
warn C000 fmf node failed schema validation
warn C001 summary key is missing
pass P001 correct keys are used
pass P002 execute step defined with "how"
pass P003 execute step methods are all known
pass P004 discover step methods are all known
skip P005 no remote fmf ids defined
pass P006 phases have unique names
pass P007 execute phase 'default-0' does not require specific guest

/containerfile_needs_deps
warn C000 key "add-deps" not recognized by schema /schemas/provision/container
warn C000 key "containerfile" not recognized by schema /schemas/provision/container
warn C000 key "containerfile-workdir" not recognized by schema /schemas/provision/container
warn C000 key "disk" not recognized by schema /schemas/provision/container
warn C000 value of "how" is not "container"
warn C000 key "add-deps" not recognized by schema /schemas/provision/mrack
warn C000 key "containerfile" not recognized by schema /schemas/provision/mrack
warn C000 key "containerfile-workdir" not recognized by schema /schemas/provision/mrack
warn C000 key "disk" not recognized by schema /schemas/provision/mrack
warn C000 value of "how" is not "beaker"
warn C000 key "add-deps" not recognized by schema /schemas/provision/virtual
warn C000 key "containerfile" not recognized by schema /schemas/provision/virtual
warn C000 key "containerfile-workdir" not recognized by schema /schemas/provision/virtual
warn C000 value of "how" is not "virtual"
warn C000 key "add-deps" not recognized by schema /schemas/provision/artemis
warn C000 key "containerfile" not recognized by schema /schemas/provision/artemis
warn C000 key "containerfile-workdir" not recognized by schema /schemas/provision/artemis
warn C000 key "disk" not recognized by schema /schemas/provision/artemis
warn C000 value of "how" is not "artemis"
warn C000 key "add-deps" not recognized by schema /schemas/provision/minute
warn C000 key "containerfile" not recognized by schema /schemas/provision/minute
warn C000 key "containerfile-workdir" not recognized by schema /schemas/provision/minute
warn C000 key "disk" not recognized by schema /schemas/provision/minute
warn C000 value of "how" is not "minute"
warn C000 key "add-deps" not recognized by schema /schemas/provision/connect
warn C000 key "containerfile" not recognized by schema /schemas/provision/connect
warn C000 key "containerfile-workdir" not recognized by schema /schemas/provision/connect
warn C000 key "disk" not recognized by schema /schemas/provision/connect
warn C000 value of "how" is not "connect"
warn C000 key "add-deps" not recognized by schema /schemas/provision/local
warn C000 key "containerfile" not recognized by schema /schemas/provision/local
warn C000 key "containerfile-workdir" not recognized by schema /schemas/provision/local
warn C000 key "disk" not recognized by schema /schemas/provision/local
warn C000 value of "how" is not "local"
warn C000 fmf node failed schema validation
warn C001 summary key is missing
pass P001 correct keys are used
pass P002 execute step defined with "how"
pass P003 execute step methods are all known
pass P004 discover step methods are all known
skip P005 no remote fmf ids defined
pass P006 phases have unique names
pass P007 execute phase 'default-0' does not require specific guest

/image_includes_deps
warn C000 key "add-deps" not recognized by schema /schemas/provision/container
warn C000 key "containerimage" not recognized by schema /schemas/provision/container
warn C000 key "disk" not recognized by schema /schemas/provision/container
warn C000 value of "how" is not "container"
warn C000 key "add-deps" not recognized by schema /schemas/provision/mrack
warn C000 key "containerimage" not recognized by schema /schemas/provision/mrack
warn C000 key "disk" not recognized by schema /schemas/provision/mrack
warn C000 value of "how" is not "beaker"
warn C000 key "add-deps" not recognized by schema /schemas/provision/virtual
warn C000 key "containerimage" not recognized by schema /schemas/provision/virtual
warn C000 value of "how" is not "virtual"
warn C000 key "add-deps" not recognized by schema /schemas/provision/artemis
warn C000 key "containerimage" not recognized by schema /schemas/provision/artemis
warn C000 key "disk" not recognized by schema /schemas/provision/artemis
warn C000 value of "how" is not "artemis"
warn C000 key "add-deps" not recognized by schema /schemas/provision/minute
warn C000 key "containerimage" not recognized by schema /schemas/provision/minute
warn C000 key "disk" not recognized by schema /schemas/provision/minute
warn C000 value of "how" is not "minute"
warn C000 key "add-deps" not recognized by schema /schemas/provision/connect
warn C000 key "containerimage" not recognized by schema /schemas/provision/connect
warn C000 key "disk" not recognized by schema /schemas/provision/connect
warn C000 value of "how" is not "connect"
warn C000 key "add-deps" not recognized by schema /schemas/provision/local
warn C000 key "containerimage" not recognized by schema /schemas/provision/local
warn C000 key "disk" not recognized by schema /schemas/provision/local
warn C000 value of "how" is not "local"
warn C000 fmf node failed schema validation
warn C001 summary key is missing
pass P001 correct keys are used
pass P002 execute step defined with "how"
pass P003 execute step methods are all known
pass P004 discover step methods are all known
skip P005 no remote fmf ids defined
pass P006 phases have unique names
pass P007 execute phase 'default-0' does not require specific guest

/image_needs_deps
warn C000 key "add-deps" not recognized by schema /schemas/provision/container
warn C000 key "containerimage" not recognized by schema /schemas/provision/container
warn C000 key "disk" not recognized by schema /schemas/provision/container
warn C000 value of "how" is not "container"
warn C000 key "add-deps" not recognized by schema /schemas/provision/mrack
warn C000 key "containerimage" not recognized by schema /schemas/provision/mrack
warn C000 key "disk" not recognized by schema /schemas/provision/mrack
warn C000 value of "how" is not "beaker"
warn C000 key "add-deps" not recognized by schema /schemas/provision/virtual
warn C000 key "containerimage" not recognized by schema /schemas/provision/virtual
warn C000 value of "how" is not "virtual"
warn C000 key "add-deps" not recognized by schema /schemas/provision/artemis
warn C000 key "containerimage" not recognized by schema /schemas/provision/artemis
warn C000 key "disk" not recognized by schema /schemas/provision/artemis
warn C000 value of "how" is not "artemis"
warn C000 key "add-deps" not recognized by schema /schemas/provision/minute
warn C000 key "containerimage" not recognized by schema /schemas/provision/minute
warn C000 key "disk" not recognized by schema /schemas/provision/minute
warn C000 value of "how" is not "minute"
warn C000 key "add-deps" not recognized by schema /schemas/provision/connect
warn C000 key "containerimage" not recognized by schema /schemas/provision/connect
warn C000 key "disk" not recognized by schema /schemas/provision/connect
warn C000 value of "how" is not "connect"
warn C000 key "add-deps" not recognized by schema /schemas/provision/local
warn C000 key "containerimage" not recognized by schema /schemas/provision/local
warn C000 key "disk" not recognized by schema /schemas/provision/local
warn C000 value of "how" is not "local"
warn C000 fmf node failed schema validation
warn C001 summary key is missing
pass P001 correct keys are used
pass P002 execute step defined with "how"
pass P003 execute step methods are all known
pass P004 discover step methods are all known
skip P005 no remote fmf ids defined
pass P006 phases have unique names
pass P007 execute phase 'default-0' does not require specific guest

    warn: /containerfile_includes_deps:provision - {'how': 'bootc', 'add-deps': False, 'containerfile': '$TMT_BOOTC_CONTAINERFILE_RUNDIR/includes_deps.containerfile', 'containerfile-workdir': '.', 'disk': 20, 'name': 'default-0'} is not valid under any of the given schemas
    warn: /containerfile_needs_deps:provision - {'how': 'bootc', 'add-deps': True, 'containerfile': '$TMT_BOOTC_CONTAINERFILE_RUNDIR/needs_deps.containerfile', 'containerfile-workdir': '.', 'disk': 20, 'name': 'default-0'} is not valid under any of the given schemas
    warn: /image_includes_deps:provision - {'how': 'bootc', 'add-deps': False, 'containerimage': 'localhost/tmt-bootc-includes-deps', 'disk': 20, 'name': 'default-0'} is not valid under any of the given schemas
    warn: /image_needs_deps:provision - {'how': 'bootc', 'add-deps': True, 'containerimage': 'localhost/tmt-bootc-needs-deps', 'disk': 20, 'name': 'default-0'} is not valid under any of the given schemas
Lint checks on all
pass G001 no duplicate ids detected
```

thrix commented:
/packit build

thrix commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,242 @@
+        """ Build the bootc disk from a container image using bootc image builder """
+        self._logger.debug("Building bootc disk image")
+        tmt.utils.Command(
+            "podman", "run", "--rm", "--privileged",
```
Whould this not run under root user for it to work? I get this error on my laptop when trying to run it:

```
❯ pwd
/var/home/mvadkert/git/github.com/teemtee/tmt/tests/provision/bootc/data
❯ tmt run -vvv -e TMT_BOOTC_CONTAINERFILE_RUNDIR=.
/var/home/mvadkert/.local/share/tmt/run-002
    warn: /containerfile_includes_deps:provision - {'how': 'bootc', 'add-deps': False, 'containerfile': '$TMT_BOOTC_CONTAINERFILE_RUNDIR/includes_deps.containerfile', 'containerfile-workdir': '.', 'disk': 20} is not valid under any of the given schemas
    warn: /containerfile_needs_deps:provision - {'how': 'bootc', 'add-deps': True, 'containerfile': '$TMT_BOOTC_CONTAINERFILE_RUNDIR/needs_deps.containerfile', 'containerfile-workdir': '.', 'disk': 20} is not valid under any of the given schemas
    warn: /image_includes_deps:provision - {'how': 'bootc', 'add-deps': False, 'containerimage': 'localhost/tmt-bootc-includes-deps', 'disk': 20} is not valid under any of the given schemas
    warn: /image_needs_deps:provision - {'how': 'bootc', 'add-deps': True, 'containerimage': 'localhost/tmt-bootc-needs-deps', 'disk': 20} is not valid under any of the given schemas
Found 4 plans.

/containerfile_includes_deps
    discover
        how: shell
        order: 50
        summary: 1 test selected
            /booted image
    provision
        queued provision.provision task #1: default-0
        
        provision.provision task #1: default-0
        how: bootc
        order: 50
        name: default-0
        how: bootc
        image: file:///var/home/mvadkert/.local/share/tmt/run-002/containerfile_includes_deps/provision/default-0/qcow2/disk.qcow2
        containerfile: ./includes_deps.containerfile
        add deps: False
        memory: 2048 MB
        disk: 20 GB
        fail: Command 'podman run --rm --privileged -v /var/lib/containers/storage:/var/lib/containers/storage --security-opt label=type:unconfined_t -v /var/home/mvadkert/.local/share/tmt/run-002/containerfile_includes_deps/provision/default-0:/output quay.io/centos-bootc/bootc-image-builder:latest build --type qcow2 --local localhost/tmtbase-1729527982' returned 1.
    finish
        warn: Nothing to finish, no guests provisioned.

plan failed

The exception was caused by 1 earlier exceptions

Cause number 1:

    provision step failed

    The exception was caused by 1 earlier exceptions

    Cause number 1:

        Command 'podman run --rm --privileged -v /var/lib/containers/storage:/var/lib/containers/storage --security-opt label=type:unconfined_t -v /var/home/mvadkert/.local/share/tmt/run-002/containerfile_includes_deps/provision/default-0:/output quay.io/centos-bootc/bootc-image-builder:latest build --type qcow2 --local localhost/tmtbase-1729527982' returned 1.

        stderr (8/8 lines)
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        Trying to pull quay.io/centos-bootc/bootc-image-builder:latest...
        Getting image source signatures
        Copying blob sha256:311336741819859e92064b81199a47397788d736d764f561a89cba9583600387
        Copying blob sha256:35056b112ca2a75be2e91b1076785d9712b1790aea49ebfd61f1799d090d9057
        Copying blob sha256:2edf720f10afc7137e2387bdb3ba800eec2602ac16da698ec19642a4d6fbc4dd
        Copying config sha256:767a10b2f63a72d479d9b102c2016790c63b045f6b9533c639fed605c299ee14
        Writing manifest to image destination
        2024/10/21 16:27:47 error: cannot validate the setup: this command must be run in rootful (not rootless) podman
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

per:

https://github.com/osbuild/bootc-image-builder?tab=readme-ov-file#building

And we will need to put it under the `--feeling-safe` option.

Personally I feel this might a bit limit the usability here, I am wondering if it would not make more sense to spin a VM and run this command in it, but yeah, that will make this a lot harder to do, but avoid the `root` requirement of this feature.

thrix replied:
We can do that later, any idea if this privileged root requirement will ever go away?

ckyrouac replied:
currently, creating the bootc disk requires a loopback mount which is why root is required. I don't think that requirement will go away anytime soon, although it's a common pain point so it might happen sooner rather than later.

> Personally I feel this might a bit limit the usability here, I am wondering if it would not make more sense to spin a VM and run this command in it

A common pattern is to run podman-machine (a VM) to enable rootful podman. Then default all podman commands to use the machine, e.g.

```
podman machine init --rootful --disk-size 200 --memory 8192 --cpus 8 -v /var/tmp/tmt:/var/tmp/tmt -v $HOME:$HOME
export CONTAINER_CONNECTION=podman-machine-default-root
podman run ...
```

ckyrouac replied:
this is documented in the PR here: https://github.com/teemtee/tmt/pull/3161/files#diff-f05fc41435a3de82fd67c56c9bb94d0b781ddaf4182fdf907b4d02ce3531476eR129

thrix replied:
@ckyrouac can you please incorporate `podman machine init` so we do not need to care about it, i.e for the time of building the image, it would be done automatically, not requiring any user action.

thrix replied:
@ckyrouac could we detect if running as root and have a default based on that? Seems like a good middle ground. And make an option to be able to force whatever if that is what the user wants?

thrix commented:
> I am getting validation errors, I assume the jsonschema needs an update for the new provision plugin:
> 
> ```
> ❯ pwd
> /var/home/mvadkert/git/github.com/teemtee/tmt/tests/provision/bootc/data
> 
> ❯ tmt lint
>     warn: /containerfile_includes_deps:provision - {'how': 'bootc', 'add-deps': False, 'containerfile': '$TMT_BOOTC_CONTAINERFILE_RUNDIR/includes_deps.containerfile', 'containerfile-workdir': '.', 'disk': 20} is not valid under any of the given schemas
>     warn: /containerfile_needs_deps:provision - {'how': 'bootc', 'add-deps': True, 'containerfile': '$TMT_BOOTC_CONTAINERFILE_RUNDIR/needs_deps.containerfile', 'containerfile-workdir': '.', 'disk': 20} is not valid under any of the given schemas
>     warn: /image_includes_deps:provision - {'how': 'bootc', 'add-deps': False, 'containerimage': 'localhost/tmt-bootc-includes-deps', 'disk': 20} is not valid under any of the given schemas
>     warn: /image_needs_deps:provision - {'how': 'bootc', 'add-deps': True, 'containerimage': 'localhost/tmt-bootc-needs-deps', 'disk': 20} is not valid under any of the given schemas
> /containerfile_includes_deps
> warn C000 key "add-deps" not recognized by schema /schemas/provision/container
> warn C000 key "containerfile" not recognized by schema /schemas/provision/container
> warn C000 key "containerfile-workdir" not recognized by schema /schemas/provision/container
> warn C000 key "disk" not recognized by schema /schemas/provision/container
> warn C000 value of "how" is not "container"
> ... snip ....
> ```

@ckyrouac FTR the respective schemas are in https://github.com/teemtee/tmt/tree/bc64a2fd569c2d1179cb63bd50dbefff429d4f14/tmt/schemas/provision, you will need to add one for this new provision plugin

happz commented:
Bumping to 1.39: we decided to shorten the 1.38 release cycle and ship it tomorrow to deliver two important features. 1.39 should be also shortened, taking just 2 weeks.

thrix commented:
@ckyrouac Let us know if you would need any help and I promise I will be more timely on the reviews. Thanks for bringing this feature to `tmt` and sorry this took so long.

ckyrouac commented:
@thrix no worries. I pushed all the requested changes in a few commits to make it easier to review. I can squash after review if you want.

martinhoyer reviewed (approved):
Very nice, thanks!  
In my testing, I've hit https://github.com/containers/podman/issues/23127 and wondering if we could print a hint for users to install `podman-machine`. Perhaps we could also add it as "Recommends" to `tmt+provision-virtual` subpackage? Not sure.

Other thing I've noticed is that bootc provisioning does not 'require' `'tmt+provision-container' ` and `tmt+provision-virtual`. i.e. it should, afaik, fail with:  
`Install 'tmt+provision-container' to provision using this method.`  
while it instead fails on:
`[Errno 2] No such file or directory: 'podman'`

happz commented:
/packit build

happz commented on tmt/schemas/provision/bootc.yaml:
```diff
@@ -0,0 +1,71 @@
+    type: string
+
+  add_deps:
+    type: bool
```
```suggestion
    type: bool
    type: boolean
```

happz commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,311 @@
+PODMAN_ENV = tmt.utils.Environment.from_dict({"CONTAINER_CONNECTION": f'{PODMAN_MACHINE_NAME}-root'})
+PODMAN_MACHINE_CPU = os.getenv('TMT_BOOTC_PODMAN_MACHINE_CPU', '2')
+PODMAN_MACHINE_MEM = os.getenv('TMT_BOOTC_PODMAN_MACHINE_MEM', '2048')
+PODMAN_MACHINE_DISK_SIZE = os.getenv('TMT_BOOTC_PODMAN_MACHINE_DISK_SIZE', '50')
```
`PODMAN_MACHINE_*` - at first I wanted to convert these into keys in `BootcData`, i.e. make them configurable via fmf and CLI. As they exist here and now it's just environment variable, and that's not consistent with other plugins and their configurables.

But then: it's just `hardware`, we already have ways how a HW requirement should be specified, https://tmt.readthedocs.io/en/stable/spec/hardware.html#cpu, and here we have pretty much the following:

```
hardware:
  cpu:
    processors: ...
  memory: ...
  disk:
    - size: ...
```

It's not really `hardware` key, rather `podman-machine-hwardware`, but the content, the way how HW requirement are given, should be the same. So, I have the following proposal:

* please, drop the environment variables from these: they do not follow the schema we established for plugin keys, https://tmt.readthedocs.io/en/stable/overview.html#plugin-variables, and their very existence suggests it should be possible to modify the number of `podman-machine` CPUs this way - it shouldn't we need to stick to HW specification we already have to remain consistent.
* we're using Pint to help us with units, so we could track the magnitude and unit together, therefore `2048` should be `2048 MB` or something similar: see how `virtual` does it, https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/testcloud.py#L180. The code below would then use `PODMAN_MACHINE_MEM.to('MB').magnitude` when passing the value to commands.
* together, we would end up with something like the following:
  ```python
  DEFAULT_PODMAN_MACHINE_CPU = 2
  DEFAULT_PODMAN_MACHINE_MEM: 'Size' = tmt.hardware.UNITS('2048 MB')
  DEFAULT_PODMAN_MACHINE_DISK_SIZE: 'Size' = tmt.hardware.UNITS('50 GB')
  ```

That would align your code with the rest of tmt, but it would also remove the configurability of these three values, which would be unfortunate, but I strongly believe we should allow it, but not this way. And I would submit a follow-up patch implementing `podman-machine-hardware` key, to open this to users - I don't want to block this PR on it, but I also don't want to let `TMT_BOOTC_PODMAN_MACHINE_MEM` being used, because once it's out there, it's much harder to get rid of it...

happz commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,311 @@
+            echo "[Install]" >> /etc/systemd/system/testcloud.service && \
+            echo "WantedBy=multi-user.target" >> /etc/systemd/system/testcloud.service && \
+            systemctl enable testcloud.service
+        '''
```
I have a feeling this template (and the one with the HTTP server above) is starting to get out of hands - a complicated multiline string buried in a function. How about making it a module member, e.g. `CONTAINERFILE_TEMPLATE`, with some nice indentation of commands in it, maybe a comment or two? It's really hard for me to track what's happening there, it's very dense. Something like this would make it easier to decipher, at least for me:

```python
CONTAINER_TEMPLATE = """
FROM {{ base_image }}

RUN    dnf -y install cloud-init rsync \
    # Make sure cloud-init runs
    && ln -s ../cloud-init.target /usr/lib/systemd/system/default.target.wants \
    # ...
    && rm /usr/local -rf \
    && ln -sr /var/usrlocal /usr/local \
    && mkdir -p /var/usrlocal/bin \
    && dnf clean all \
    # Workaround some testcloud problem??
    && echo "{{ testcloud_guest }}" >> /opt/testcloud-guest.sh \
    ...
```

Note I moved `&&` to the beginning of lines - I'm used to this way, but if you believe `&&` at the end is better, no problem. I'd just keep commands spread across multiple lines, with comments and better visibility into why they are there happening - as I reviewer, I can see some manipulation with `/usr/local`, but that's all... A hint or two would help future readers.

happz commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,311 @@
+            "--disk-size", PODMAN_MACHINE_DISK_SIZE,
+            "--memory", PODMAN_MACHINE_MEM,
+            "--cpus", PODMAN_MACHINE_CPU,
+            "-v", "/var/tmp/tmt:/var/tmp/tmt",
```
I'd make this path module-level member, some `DEFAULT_TMP_PATH` or something - the issue is, it's in the middle of code, it's not configurable here, it looks like a workdir root users **can** and often **do** override, so it will need to be configurable in the future, and linters complain it's an unsafe use of a temporary directory. It's not, it's perfectly fine, but moving it to module level, we could waive silence linter more easily.

ckyrouac commented:
force pushed updates addressing all the suggestions.

I had broken out the podman-machine cpu/mem parameters into environment variables with the idea that the podman-machine would be spun up once for multiple tests. With the current code, a new machine is created and destroyed for each test run. That's mostly because I'm not sure where/how to hook into the higher level run code. Although, the more I think about it, I guess allowing each test to specify the podman-machine cpu/mem requirements makes sense because some tests might require significantly more resources than other tests.

> please, install and enable pre-commit in your repo

hmm, I have it installed but it didn't run on one of those commits for some reason.

psss commented:
/packit build

ckyrouac commented:
hmm, looking into the test failures

thrix commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,364 @@
+touch /etc/environment && \
+echo "export PATH=$PATH:/var/lib/tmt/scripts" >> /etc/environment && \
+
+# create a simple webserver used by tmt to verify the VM is running
```
@frantisekz @ckyrouac I wonder if this is needed now that testcloud was fixed to be able to start this sever by itself:

https://github.com/teemtee/testcloud/pull/6/files

ckyrouac replied:
nice! Removed it

ckyrouac commented:
not sure why the CI tests were failing. The tests are passing locally, in CI the setup was failing to create the /var/tmp/tmt/test-XXXX directory. I can't find the previous run results to get the exact error after I pushed the latest updates.

psss commented:
One of the old jobs is [here](https://artifacts.dev.testing-farm.io/3b72dcfa-de4e-4a1a-8323-02ebdcb7c88c/). There seems to be a problem with `tmt init` at the beginning:

```
:: [ 09:08:32 ] :: [  BEGIN   ] :: Running 'tmt init'
Path '/var/ARTIFACTS/work-bootcdy90p5cu/plans/provision/bootc/discover/default-0/tests/tests/provision/bootc' already has a parent fmf tree root '/var/ARTIFACTS/work-bootcdy90p5cu/plans/provision/bootc/discover/default-0/tests/.fmf'.
Do you really want to initialize a nested tree? [y/N]: Aborted!
:: [ 09:08:33 ] :: [   FAIL   ] :: Command 'tmt init' (Expected 0, got 1)
```

Using `tmt init --force` should help.

ckyrouac commented:
ah, thanks for the link. The latest commit should fix it. Not sure if you need to trigger another test run or not.

psss commented:
/packit build

psss commented on tmt.spec:
```diff
@@ -90,6 +90,20 @@ Recommends:     qemu-system-x86-core
+Recommends:     virtiofsd
+
+%description -n tmt+provision-bootc %_metapackage_description
+
```
At least a minimal `%files` section is also needed to convince `rpm-build` to actually create the subpackage. See bottom of the spec.

ckyrouac commented:
/packit build

happz commented:
/packit build

happz commented on tmt.spec:
```diff
@@ -90,6 +90,20 @@ Recommends:     qemu-system-x86-core
 
 %description -n tmt+provision-virtual %_metapackage_description
 
+%package -n     tmt+provision-bootc
```
RPM build fails with the following:

```
Processing files: tmt+provision-bootc-1.39.dev888-1.20241106153118869510.pr3161.23.g7d63c518.fc39.noarch

Error: The package name contains an extras name `provision-bootc` that was not found in the metadata.
Check if the extras were removed from the project. If so, consider removing the subpackage and obsoleting it from another.

error: Dependency tokens must begin with alpha-numeric, '_' or '/': *** PYTHON_EXTRAS_NOT_FOUND_ERROR___SEE_STDERR ***

Error: The package name contains an extras name `provision-bootc` that was not found in the metadata.
Check if the extras were removed from the project. If so, consider removing the subpackage and obsoleting it from another.

error: Dependency tokens must begin with alpha-numeric, '_' or '/': *** PYTHON_EXTRAS_NOT_FOUND_ERROR___SEE_STDERR ***
```

I'm not an expert on rpmbuild, but IIUIC, the problem is no `provision-bootc` extras in `pyproject.toml`. Even an empty one, like `provision-container`, could resolve the problem: https://github.com/teemtee/tmt/blob/main/pyproject.toml#L62

ckyrouac replied:
thanks, pushed the change to pyproject.tml

happz commented:
/packit build

ckyrouac commented:
The tests are failing with `Failed to connect socket to '/var/run/libvirt/virtqemud-sock': No such file or directory`. Any ideas why?

omertuc commented:
> The tests are failing with `Failed to connect socket to '/var/run/libvirt/virtqemud-sock': No such file or directory`. Any ideas why?

I ran into this before locally, I solved it with `systemctl enable virtqemud.socket`

happz commented:
> The tests are failing with `Failed to connect socket to '/var/run/libvirt/virtqemud-sock': No such file or directory`. Any ideas why?

Tests depend on additional setup, shouldn't we add a dedicated plan & let it run in the `provision` check? IIUIC, that's what we did with `provision-virtual` tests, they are now grouped under `/plans/provision/virtual` which tasks care of the necessary setup, starting libvirt is exactly one of the steps in `plans/provision/virtual.fmf`, and since bootc provisioning plugin depends on an extra setup as well, we should isolate them in the same manner.

@psss does it make sense? It does look like another provision-foo plan is needed, and I think we can add it to the same Github check as `virtual` tests.

ckyrouac commented:
@happz added a prepare step to the bootc plan. Could you trigger another test run?

psss commented:
/packit build

psss commented:
> @psss does it make sense? It does look like another provision-foo plan is needed, and I think we can add it to the same Github check as `virtual` tests.

Yes, this definitely makes sense. Let's have this handled in a consistent way.

ckyrouac commented:
If I understand this right, I believe this was already in place. The [prepare step](https://github.com/teemtee/tmt/pull/3161/files#diff-93304e094886795c4b6e58cf3826294d1ba36e115721a851bd284cfdc31510bfR10) just wasn’t in place yet but is now.

Now the bootc tests pass but there is an [avc failure](https://artifacts.dev.testing-farm.io/f59d5356-f9a9-400a-b3c7-a9ef75602018/work-bootchri1stdf/plans/provision/bootc/execute/data/guest/default-0/tests/provision/bootc-1/checks/avc.txt) that I don’t understand. Any ideas?

happz commented:
> If I understand this right, I believe this was already in place. The [prepare step](https://github.com/teemtee/tmt/pull/3161/files#diff-93304e094886795c4b6e58cf3826294d1ba36e115721a851bd284cfdc31510bfR10) just wasn’t in place yet but is now.
> 
> Now the bootc tests pass but there is an [avc failure](https://artifacts.dev.testing-farm.io/f59d5356-f9a9-400a-b3c7-a9ef75602018/work-bootchri1stdf/plans/provision/bootc/execute/data/guest/default-0/tests/provision/bootc-1/checks/avc.txt) that I don’t understand. Any ideas?

A rebase should help, the problem has been addressed in https://github.com/teemtee/tmt/pull/3355

happz commented:
/packit build

martinhoyer commented:
> If I understand this right, I believe this was already in place. The [prepare step](https://github.com/teemtee/tmt/pull/3161/files#diff-93304e094886795c4b6e58cf3826294d1ba36e115721a851bd284cfdc31510bfR10) just wasn’t in place yet but is now.
> 
> Now the bootc tests pass but there is an [avc failure](https://artifacts.dev.testing-farm.io/f59d5356-f9a9-400a-b3c7-a9ef75602018/work-bootchri1stdf/plans/provision/bootc/execute/data/guest/default-0/tests/provision/bootc-1/checks/avc.txt) that I don’t understand. Any ideas?

I'm guessing it's the usage of /var, nosuid transition:
https://danwalsh.livejournal.com/78312.html

I can see `container-selinux` package being installed in the log btw. Maybe a bug there on in podman-machine package? Not sure how to solve this. @psss @happz

happz commented:
> > If I understand this right, I believe this was already in place. The [prepare step](https://github.com/teemtee/tmt/pull/3161/files#diff-93304e094886795c4b6e58cf3826294d1ba36e115721a851bd284cfdc31510bfR10) just wasn’t in place yet but is now.
> > Now the bootc tests pass but there is an [avc failure](https://artifacts.dev.testing-farm.io/f59d5356-f9a9-400a-b3c7-a9ef75602018/work-bootchri1stdf/plans/provision/bootc/execute/data/guest/default-0/tests/provision/bootc-1/checks/avc.txt) that I don’t understand. Any ideas?
> 
> I'm guessing it's the usage of /var, nosuid transition: https://danwalsh.livejournal.com/78312.html
> 
> I can see `container-selinux` package being installed in the log btw. Maybe a bug there on in podman-machine package? Not sure how to solve this. @psss @happz

It has been reported for bootc image builder, see https://github.com/osbuild/bootc-image-builder/issues/645. It does look like something bootc either should not do, or ask selinux-policy maintainers to allow.

martinhoyer commented:
> Hmm that probably relates to our initialization of containers/storage at bootc install time. But I am confused as to why we may only be seeing this here. It may relate somehow to the host policy? We do have some coverage for this in other CI tests… Long story short if this theory is right then we need to suppress the domain transition or maybe just switch to skopeo which we probably want to do anyways.

Thanks,
@cgwalters @ckyrouac Would you prefer merging it "as is", perhaps like a tech preview or addressing the selinux issue first, most likely sliding to the next release cycle?

ckyrouac commented:
I will be afk on paternity leave soon (most likely in a week), so I won't be able to work on this much for a few months. So, unless someone else can pick up the work, I think merging as-is makes the most sense.

martinhoyer commented:
> I will be afk on paternity leave soon (most likely in a week)  

Congratulations in advance sir!

psss commented:
> So, unless someone else can pick up the work, I think merging as-is makes the most sense.

Ok, then I suggest to mark the `avc` failure as expected for now and include this in `1.39` as a tech preview. I've added a short release not in 44b0cd38. Does it look ok?

psss commented on plans/provision/bootc.fmf:
```diff
@@ -0,0 +1,39 @@
+summary: Bootc virtual machine via testcloud
```
This plan is currently run under the `full` packit job. Seems to be working fine, but this means that it is executed for all targets. Should we move it under the `provision` packit job? @happz, I guess that's what you meant in your [comment](https://github.com/teemtee/tmt/pull/3161#issuecomment-2468435410)? Or are we good as it is now?

psss replied:
Covered by f3772f6533398dcd93b28f6c3f88bed83597e9a4.

psss commented on tests/provision/bootc/test.sh:
```diff
@@ -0,0 +1,64 @@
+    rlPhaseStartTest "Image that already includes dependencies"
+        rlRun "podman build . -f $CONTAINERFILE_INCLUDES_DEPS -t $IMAGE_INCLUDES_DEPS"
+        rlRun "cp $IMAGE_INCLUDES_DEPS_PLAN ."
+        rlRun "tmt -vvvvv run -i $run"
```
If reusing the same run `--id` it's better to use `tmt run --scratch` to remove previous run content before executing a new run. Otherwise this results into following tmt runs to just report `done`.

psss replied:
Should be covered by f781e34b.

psss commented on tests/provision/bootc/test.sh:
```diff
@@ -0,0 +1,64 @@
+rlJournalStart
+    rlPhaseStartSetup
+        # cleanup previous runs
+        test -d /var/tmp/tmt/testcloud && rlRun "rm -rf /var/tmp/tmt/testcloud"
```
If we need to do a complete cleanup, what about using `TMT_WORKDIR_ROOT` instead to choose a different workdir root? We're trying to write tests so that they can be easily/safely executed on user laptop as well. This would remove for example all fetched images.

psss commented on tmt.spec:
```diff
@@ -90,6 +90,20 @@ Recommends:     qemu-system-x86-core
+Provides:       tmt-provision-bootc == %{version}-%{release}
+%if 0%{?fedora} < 40
+Obsoletes:      tmt-provision-bootc < %{version}-%{release}
+%endif
```
There was no `tmt-provision-bootc` package before, so I think we can drop the `Obsoletes` here.

psss commented on tmt/steps/provision/bootc.py:
```diff
@@ -0,0 +1,348 @@
+                ).run(cwd=self.workdir, stream_output=True, logger=self._logger)
+        except BaseException:
+            self._logger.debug(
+                "Unable to remove podman machine {PODMAN_MACHINE_NAME}, it might not exist")
```
```suggestion
                "Unable to remove podman machine '{PODMAN_MACHINE_NAME}', it might not exist.")
```

psss reviewed (changes requested):
Added a couple of comments. I tried the minimal example but ended up with the following error:

```
Command 'podman machine start podman-machine-tmt' returned 125.

# stdout (1/1 lines)
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Starting machine "podman-machine-tmt"
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# stderr (2/2 lines)
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
time="2024-11-15T14:33:33+01:00" level=error msg="process 419571 has not ended"
Error: failed to find virtiofsd: exec: "virtiofsd": executable file not found in $PATH
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I tried this under a regular user. Do I understand it correctly that it is necessary to always run this under root?

happz commented:
/packit build

ckyrouac commented:
@psss I addressed most of your suggestions in recent pushes. I probably won’t have time to refactor the tests per your suggestions though.

@happz I think I removed your commit via a force push, sorry. I’m done pushing now :-)

psss commented:
/packit build

psss commented:
/packit build

psss commented:
I've added a bunch of changes in f781e34b, mainly to fix & simplify tests:

* Make sure that all test cases are actually executed.
* Run tests directly in the `data` directory.
* Simplify plans using inheritance.
* Plus some minor adjustments.

Last remaining problem is the avc check. I guess we should just disable it if we cannot `xfail` it?

psss commented:
/packit build

psss commented:
/packit build

psss commented:
/packit build

psss commented:
/packit build

psss commented:
/packit build
